### PR TITLE
Add support for Range and RangeInclusive

### PR DESCRIPTION
**Pull request by dvdplm**
Adding support for `Range` and `RangeInclusive` (added in https://github.com/paritytech/parity-scale-codec/pull/269 to `parity-scale-codec` and to polkadot.js in https://github.com/polkadot-js/api/pull/3791).

Closes https://github.com/paritytech/scale-info/issues/123

### CODE REVIEW

File: src/impls.rs [hunk 1]
```diff
@@ -347,6 +351,26 @@ where
...
+    }
+}
+
+impl<Idx> TypeInfo for RangeInclusive<Idx>
+where
+    Idx: TypeInfo + 'static + PartialOrd + core::fmt::Debug,
+{
+    type Identity = Self;
+    fn type_info() -> Type {
+        TypeDefRangeInclusive::new::<Idx>().into()
```
[the remark this answers is not in the document]
Comment (ascjones) [reply src/impls.rs:372]: Could add an `inclusive: bool` field? Then we would no longer need the two top level variants.

File: src/ty/mod.rs [hunk 2]
```diff
@@ -411,6 +415,50 @@ where
...
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Debug)]
+pub struct TypeDefRange<T: Form = MetaForm> {
+    start: T::Type,
+    end: T::Type,
+    inclusive: bool,
+}
+
+impl TypeDefRange {
+    /// Creates a new [`TypeDefRange`] for the supplied index type. Use the `inclusive` parameter to
+    /// distinguish between [`Range`] and [`RangeInclusive`] range types.
```
Comment (ascjones) [inline src/ty/mod.rs:437]: I don't think these links will resolve because `Range` and `RangeInclusive` are not in scope here?

File: src/impls.rs [hunk 3]
```diff
@@ -347,6 +352,26 @@ where
     }
 }
 
+impl<Idx> TypeInfo for Range<Idx>
+where
+    Idx: TypeInfo + 'static + PartialOrd + core::fmt::Debug,
```
Comment (ascjones) [inline src/impls.rs:357]: We should be consistent and use `fmt::Debug` from the prelude, also for other occurrences of `core::fmt::Debug`

**Comment by ascjones** [comment]
> Fixed the ui tests in #125

Merged this, so can update this branch now

File: src/ty/mod.rs (hunk 2 above)
Comment (dvdplm) [reply src/ty/mod.rs:437]: good pt!

File: test_suite/tests/json.rs [hunk 4]
```diff
@@ -360,6 +364,27 @@ fn test_enum() {
...
+        closed_range: RangeInclusive<u64>,
+    }
+
+    assert_json_for_type::<TypeWithRanges>(json!({
+        "path": ["json", "TypeWithRanges"],
+        "def": {
+            "composite": {
+                "fields": [
+                    { "name": "open_range", "type": 0, "typeName": "Range<u8>" },
+                    { "name": "closed_range", "type": 2, "typeName": "RangeInclusive<u64>" },
```
Comment (ascjones) [inline test_suite/tests/json.rs:381]: I think this test should also check that the `inclusive` field is set correctly on the respective types. Might be easiest by just comparing the whole serialized type registry which we do in some other tests.

**Review by ascjones — APPROVED** [review]
Apart from the nitpick about the test, LGTM!

File: test_suite/tests/json.rs [hunk 5]
```diff
@@ -372,17 +372,56 @@ fn test_ranges() {
...
+                    "def": { "primitive": "u64" },
+                }
+            }
+        ],
+    });
+
+    let registry: PortableRegistry = registry.into();
+    assert_eq!(serde_json::to_value(registry).unwrap(), expected);
+
+    // assert_json_for_type::<TypeWithRanges>();
```
Comment (ascjones) [inline test_suite/tests/json.rs:424]: remove commented out

**Pull request merged**